## Q1 - Heap을 사용한 생일 정렬

생일 데이터를 교재 코드(heap.py)를 이용해 힙으로 저장하고, 생일이 느린 순서로 
10명의 친구를 출력하는 코드를 작성한다. 실행 결과가 셀에 나타나야 한다. 

### 코드

In [33]:
import csv
from heap import Heap

birth_heap = Heap()

with open('DS_Birthdaydata.csv', 'r', encoding='cp949') as file:
    reader = csv.DictReader(file)
    for row in reader:
        name = row['이름']
        birthday = row['생년월일'].replace('-', '').replace('.', '') 
        if birthday.isdigit():
            birth_heap.insert((int(birthday), name))

print("생일이 늦은 친구 10명:")
for _ in range(10):
    if not birth_heap.isEmpty():
        birthday, name = birth_heap.deleteMax()
        print(f"{name} - {str(birthday)}")

생일이 늦은 친구 10명:
홍서연 - 20241282
신수민 - 20051230
이서영 - 20051225
강민주 - 20051214
김민경 - 20051202
이서영 - 20051112
배시은 - 20051102
김여원 - 20051031
이서진 - 20051028
서홍빈 - 20051024


### 실행결과


생일이 늦은 친구 10명:  
홍서연 - 20241282  
신수민 - 20051230  
이서영 - 20051225  
강민주 - 20051214  
김민경 - 20051202  
이서영 - 20051112  
배시윤 - 20051102  
김여원 - 20051031  
이서진 - 20051028  
서홍빈 - 20051024  

## Q2 - Circular Doubly Linked List를 사용한 조원 생일 필터링

생일 데이터를 교재 코드(circularDoublyLinkedList.py)를 이용해 리스트로 
저장하고, 같은 조(지난 과제 지정 조원 참조)의 친구들만 이름과 생년월일을 
출력하는 코드를 작성한다. 실행 결과가 셀에 나타나야 한다. 

### 코드

In [34]:
import csv
from circularDoublyLinkedList import CircularDoublyLinkedList

my_group = {
    ("권보은", "******14"),
    ("김승연", "******30"),
    ("이서영", "******42"),
    ("이아현", "******24"),
    ("임성민", "******08"),
    ("은유빈", "******38"),
    ("이예은", "******81"),
    ("정예은", "******28"),
    ("김주원", "******43"),
    ("서홍빈", "******64")
}

birthday_list = CircularDoublyLinkedList()

with open('DS_Birthdaydata.csv', 'r', encoding='cp949') as file:
    reader = csv.DictReader(file)
    for row in reader:
        name = row['이름']
        student_id = row['학번']
        birthday = row['생년월일']
        birthday_list.append((name, student_id, birthday))

print("같은 조 친구들의 생일 정보:")
for i in range(birthday_list.size()):
    name, student_id, birthday = birthday_list.get(i)
    if (name, student_id) in my_group:
        print(f"{name} - {birthday}")



같은 조 친구들의 생일 정보:
권보은 - 20041004
김승연 - 20030124
김주원 - 20030110
서홍빈 - 20051024
은유빈 - 20040503
이서영 - 20051225
이아현 - 20010904
이예은 - 20030109
임성민 - 20021213
정예은 - 


### 실행결과

같은 조 친구들의 생일 정보:  
권보은 - 20041004  
김승연 - 20030124  
김주원 - 20030110  
서홍빈 - 20051024  
은유빈 - 20040503  
이서영 - 20051225  
이아현 - 20010904  
이예은 - 20030109  
임성민 - 20021213  
정예은 -  

## Q3 - 우선순위 큐 연습문제 풀이

교재 8장 우선 순위 큐 연습문제를 모두 푼다.  

## 문제 1

답: 아니요

이유? 최대 힙에서는 항상 루트가 가장 큰 원소 이므로, 트리의 깊은 곳에 있는 노드는 루트보다 작거나 같다.

## 문제 2

답 : A[n-1]은 항상 작은 값을 갖지 않는다.

이유? 'A[0]'은 항상 최대 힙에서 가장 큰 값을 갖는 반면, 'A[n-1]'은 단지 배열의 마지막에 있을 뿐 가장 작은 값을 갖는다고 보장 할 수 없다.

## 문제3

답: n/2개

이유? buildHeap()은 배열의 중간 인덱스부터 루트까지 스며내리기를 수행한다. 따라서 절반 정도의 원소는 스며내리기를 하지 않는다.

## 문제4

답:O(n)

이유? 각 노드의 스며내리기 연산은 깊이에 따라 짧아지기 때문에 총 시간은 O(n)이 된다.

## 문제5

답 : 아니다

이유? 힙에서 삭제 연산은 루트 노드를 제거한 뒤 마지막 원소를 루트로 옮기고, 다시 스며내리기를 수행해야하기 때문에 단순히 마지막 원소를 제거하는 것은 힙 삭제가 아니다.

## 문제6

답: 아래에서 위로 진행하는 것이 더 효율적이다.

이유? 아래에서 위로 스며내리기를 하면 자식들이 먼저 정리되어 상위 노드를 처리할 때 연산이 줄어든다. 반대로 위에서 아래로 하면 정렬되지 않은 자식을 여러번 정리해야 해서 비효율적이다.

## 문제7

답 : 스며오르기를 수행한다. 시간 복잡도는 O(log n)

이유? 값이 증가한 노드는 부모와 비교하여 더 크면 자리를 바꾸기를 한다. 이 과정을 루트까지 반복하면 힙 조건을 유지할 수 있고, 높이는 log n이므로 시간 복잡도는 O(log n)이 된다.

## Q4 - LeetCode 703. Kth Largest Element in a Stream

`KthLargest` 클래스를 정의하고, 스트림에 요소가 들어올 때마다 k번째로 큰 수를 리턴하는 구조를 구현합니다.

In [35]:
import heapq

class KthLargest:
    def __init__(self, k, nums):  
        self.k = k
        self.minheap = nums
        heapq.heapify(self.minheap)
        while len(self.minheap) > k:
            heapq.heappop(self.minheap)

    def add(self, val):
        heapq.heappush(self.minheap, val)
        if len(self.minheap) > self.k:
            heapq.heappop(self.minheap)
        return self.minheap[0]
